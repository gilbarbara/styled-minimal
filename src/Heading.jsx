import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { px, spacer, themeGet } from './utils/helpers';
import { baseStyles, headingPropTypes } from './utils/system';

import Box from './Box';

const styles = (props: Object): string => {
  const { gutterBottom } = props;
  const headingWeight = themeGet(props, 'headingWeight');
  const headingSize = themeGet(props, 'headingSizes', { key: ['size', 'as'], base: 'h1' });

  return css`
    font-size: ${px(headingSize)};
    font-family: inherit;
    font-weight: ${headingWeight};
    line-height: ${baseStyles.lineHeight};
    margin: ${spacer(3)} 0 ${gutterBottom ? spacer(3) : 0};
    
    &:first-child {
      margin-top: 0;
    }
  `;
};

export const Heading = styled(Box)`
  ${styles};
`;

Heading.displayName = 'Heading';

Heading.propTypes = {
  as: headingPropTypes,
  children: PropTypes.node.isRequired,
  gutterBottom: PropTypes.bool,
  /** element size */
  size: headingPropTypes,
  ...Box.propTypes,
};

Heading.defaultProps = {
  as: 'h1',
};

export default Heading;
