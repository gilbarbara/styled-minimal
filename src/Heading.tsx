import * as React from 'react';
import styled, { css } from 'styled-components';

import { getTheme, px, spacer } from './utils/helpers';
import { baseStyles, sx, system } from './utils/system';

import { HeadingProps } from './types';

export const StyledHeading = styled('h1')<HeadingProps>(
  props => {
    const { as, level, gutter } = props;
    const headingWeight = getTheme(props, 'headingWeight');
    const marginBottom = gutter ? spacer(3) : 0;
    let headingSize = getTheme(props, 'headingSizes')[as!];

    if (level) {
      headingSize = getTheme(props, 'headingSizes')[`h${level}`];
    }

    return css`
      ${baseStyles};
      font-size: ${px(headingSize)};
      font-family: inherit;
      font-weight: ${headingWeight};
      line-height: 1;
      margin-bottom: ${marginBottom};
      margin-top: ${spacer(2)};

      &:first-child {
        margin-top: ${0};
      }
    `;
  },
  sx,
  system,
);

export const Heading = React.forwardRef<HTMLHeadingElement, HeadingProps>((props, ref) => (
  <StyledHeading ref={ref} {...props} />
));

Heading.defaultProps = {
  as: 'h1',
  gutter: false,
};
