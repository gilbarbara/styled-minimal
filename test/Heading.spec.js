import React from 'react';
import { mount } from 'enzyme';

import Heading from '../src/Heading';

describe('Heading', () => {
  const wrapper = mount(
    <Heading>Far far away</Heading>
  );

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
