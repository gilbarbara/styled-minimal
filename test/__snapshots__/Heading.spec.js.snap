// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading should render properly 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 40px;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.4;
  margin-bottom: 0;
  margin-top: 16px;
}

.c0:first-child {
  margin-top: 0;
}

<Heading
  as="h1"
>
  <h1
    className="c0"
  >
    Far far away
  </h1>
</Heading>
`;
